kernelci/cli/base.py

```diff
@@ -560,11 +580,18 @@ def get(self, option, as_list=False):
             section = ':'.join([section_name, section_config])
         else:
             section = self._section
-        if not self._settings.has_option(section, option):
-            return None
-        value = self._settings.get(section, option).split()
-        if not as_list and len(value) == 1:
-            value = value[0]
+        if self._deprecated_settings:
+            if not self._settings.has_option(section, option):
+                return None
+            value = self._settings.get(section, option).split()
+            if not as_list and len(value) == 1:
+                value = value[0]
+        else:
+            value = None
+            if self._settings.get('DEFAULT'):
+                value = self._settings.get('DEFAULT').get(option)
+            if self._settings.get(section):
+                value = self._settings.get(section).get(option) or value
```

Comment:
BTW I think this can be optimised a bit by first looking for the value in the designated `section` as in this case there's no need to look up the `DEFAULT` value.  Also we can keep a shortcut to the default section in the object.  Something like this:

```python
# in __init__():
        if path and os.path.exists(path):
            if self._deprecated_settings:
                self._settings.read(path)
            else:
                self._settings = toml.load(path)
                self._default_section = self._settings.get('DEFAULT')

# [...]

# in get_from_section()
            value = None
            section_data = self._settings.get(section)
            if section_data:
                value = section_data.get(option)
            if value is None and self._default_section:
                value = self._default_section.get(option)
```

Reply:
So, at first, I added code to look up values from the designated section and then the default one. But then I ended up in so many `if` conditions somehow and `pycodestyle` was not happy about it.
OK. this seems fair and I'll update it.

Reply:
Oh, I remember now.
If we go with checking the specific section first and then the default one, the value from default would override.
That was the reason.

Reply:
If we define `default_section`, `self._default_section.get(option)` apparently calls `Options.get` method and it ends up in a recursive call. Hence, discarded it.

```
  File "/home/runner/work/kernelci-core/kernelci-core/kernelci/cli/base.py", line 552, in __getattr__
    return self.get(name)
  File "/home/runner/work/kernelci-core/kernelci-core/kernelci/cli/base.py", line 595, in get
    if value is None and self._default_section:
  File "/home/runner/work/kernelci-core/kernelci-core/kernelci/cli/base.py", line 552, in __getattr__
    return self.get(name)
  File "/home/runner/work/kernelci-core/kernelci-core/kernelci/cli/base.py", line 574, in get
    opt_data = self._command.get_arg_data(option)
  File "/home/runner/work/kernelci-core/kernelci-core/kernelci/cli/base.py", line 470, in get_arg_data
    return self._args_dict.get(arg_name)
RecursionError: maximum recursion depth exceeded while calling a Python object
```

Reply:
OK it would be good to get to the bottom of this and understand why it's a problem in the constructor but not in the method.  I'll take a look, it's not a blocker for this PR.

Reply:
This works for me:
```diff
diff --git a/kernelci/cli/base.py b/kernelci/cli/base.py
index 49017c8051e8..3128ff74354a 100644
--- a/kernelci/cli/base.py
+++ b/kernelci/cli/base.py
@@ -542,6 +542,7 @@ instead.", file=sys.stderr)
                 self._settings.read(path)
             else:
                 self._settings = toml.load(path)
+                self._default_section = self._settings.get('DEFAULT')
 
         self._command = command
         self._cli_args = cli_args
@@ -591,8 +592,8 @@ instead.", file=sys.stderr)
             section_data = self._settings.get(section)
             if section_data:
                 value = section_data.get(option)
-            if value is None and self._settings.get('DEFAULT'):
-                value = self._settings.get('DEFAULT').get(option)
+            if value is None and self._default_section:
+                value = self._default_section.get(option)
         return value
 
     def get_missing_args(self):
```
I guess we should also add some unit tests to cover the settings / `Options` logic at some point.

Reply:
> This works for me:
> 
> ```diff
> diff --git a/kernelci/cli/base.py b/kernelci/cli/base.py
> index 49017c8051e8..3128ff74354a 100644
> --- a/kernelci/cli/base.py
> +++ b/kernelci/cli/base.py
> @@ -542,6 +542,7 @@ instead.", file=sys.stderr)
>                  self._settings.read(path)
>              else:
>                  self._settings = toml.load(path)
> +                self._default_section = self._settings.get('DEFAULT')
>  
>          self._command = command
>          self._cli_args = cli_args
> @@ -591,8 +592,8 @@ instead.", file=sys.stderr)
>              section_data = self._settings.get(section)
>              if section_data:
>                  value = section_data.get(option)
> -            if value is None and self._settings.get('DEFAULT'):
> -                value = self._settings.get('DEFAULT').get(option)
> +            if value is None and self._default_section:
> +                value = self._default_section.get(option)
>          return value
>  
>      def get_missing_args(self):
> ```
> 
> I guess we should also add some unit tests to cover the settings / `Options` logic at some point.

Please see https://github.com/kernelci/kernelci-core/actions/runs/5140867870/jobs/9252740078

Reply:
OK this can be fixed somewhat by relying on the convention that settings don't start with `_`, that way private attributes will be found in the object while other attributes will return the settings value.

If the attribute way isn't actually used in the code then we might simply drop it and request all the code to use `.get()`, but since it's been there for a while I'm not entirely sure of this.  Maybe it's something we could consider deprecating when transitioning to TOML only, if it's actually a problem to override `__getattr__` and not that useful in practice.

Reply:
This seems like a hack to me than a fix.
Or we can use `self._deprecated_settings` flag to bifurcate both the logics.

Reply:
Found the actual problem, the `._default_section` wasn't created when the settings file was not present.  So a fix for this would be:
```diff
diff --git a/kernelci/cli/base.py b/kernelci/cli/base.py
index 49017c8051e8..28d6aa18a1c3 100644
--- a/kernelci/cli/base.py
+++ b/kernelci/cli/base.py
@@ -542,6 +542,7 @@ instead.", file=sys.stderr)
                 self._settings.read(path)
             else:
                 self._settings = toml.load(path)
+        self._default_section = self._settings.get('DEFAULT')
 
         self._command = command
         self._cli_args = cli_args
```
Then we can use `self._default_section` without causing an infinite recursion.

Reply:
> This seems like a hack to me than a fix. Or we can use `self._deprecated_settings` flag to bifurcate both the logics.

I was referring to this change by this comment:
```
+        if name.startswith('_'):
+            return super().__getattr__(name)
```

Reply:
> Found the actual problem, the `._default_section` wasn't created when the settings file was not present. So a fix for this would be:
> 
> ```diff
> diff --git a/kernelci/cli/base.py b/kernelci/cli/base.py
> index 49017c8051e8..28d6aa18a1c3 100644
> --- a/kernelci/cli/base.py
> +++ b/kernelci/cli/base.py
> @@ -542,6 +542,7 @@ instead.", file=sys.stderr)
>                  self._settings.read(path)
>              else:
>                  self._settings = toml.load(path)
> +        self._default_section = self._settings.get('DEFAULT')
>  
>          self._command = command
>          self._cli_args = cli_args
> ```
> 
> Then we can use `self._default_section` without causing an infinite recursion.

Oh, okay.
We should have a follow-up PR for this.